### PR TITLE
Improve performance of ActorInfo.TraitsInConstructOrder.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Avoid redundant enumerations of the more and unresolved enumerables. This gives an approx 3x speedup for this function.

### CODE REVIEW

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 1]
```diff
@@ -115,27 +115,34 @@ public IEnumerable<TraitInfo> TraitsInConstructOrder()
...
 				u.OptionalDependencies.All(d => // To be resolvable, all optional dependencies must be satisfied according to the following condition:
-					!unresolved.Any(u1 => testResolve(d, u1.Type)))); // All matching traits that meet this optional dependencies must be resolved first.
+					!unresolved.Any(u1 => TestResolve(d, u1.Type)))); // All matching traits that meet this optional dependencies must be resolved first.
 
 			// Continue resolving traits as long as possible.
 			// Each time we resolve some traits, this means dependencies for other traits may then be possible to satisfy in the next pass.
-			while (more.Any())
-				resolved.AddRange(more);
+			var readyToResolve = more.ToList();
+			while (readyToResolve.Any())
```
Comment (PunkPun) [inline OpenRA.Game/GameRules/ActorInfo.cs]:
```suggestion
			while (readyToResolve.Count != 0)
```

We had chosen to use this style

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 2]
```diff
@@ -115,27 +115,34 @@ public IEnumerable<TraitInfo> TraitsInConstructOrder()
 			}).ToList();
 
 			var resolved = source.Where(s => s.Dependencies.Count == 0 && s.OptionalDependencies.Count == 0).ToList();
-			var unresolved = source.Except(resolved);
+			var unresolved = source.ToHashSet();
+			unresolved.ExceptWith(resolved);
 
-			var testResolve = new Func<Type, Type, bool>((a, b) => a == b || a.IsAssignableFrom(b));
+			bool TestResolve(Type a, Type b) => a.IsAssignableFrom(b);
```
Comment (PunkPun) [inline OpenRA.Game/GameRules/ActorInfo.cs]:
```suggestion
			bool TestResolve(Type a, Type b) => a == b || a.IsAssignableFrom(b);
```

why was this removed?

Comment (RoosterDragon) [reply OpenRA.Game/GameRules/ActorInfo.cs]: `IsAssignableFrom` [does the same check](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Type.Helpers.cs,ba0cffea035fe210).

File: OpenRA.Game/GameRules/ActorInfo.cs (hunk 1 above)
Comment (RoosterDragon) [reply OpenRA.Game/GameRules/ActorInfo.cs]: Updated

File: OpenRA.Game/GameRules/ActorInfo.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Game/GameRules/ActorInfo.cs]: ok

**Review by PunkPun — DISMISSED** [review]

File: OpenRA.Game/GameRules/ActorInfo.cs (hunk 2 above)
Comment (abcdefg30) [inline OpenRA.Game/GameRules/ActorInfo.cs]: Bonus points:
```suggestion
			bool AreResolvable(Type a, Type b) => a.IsAssignableFrom(b);
```
(Which seems to me what the method does.)

**Review by abcdefg30 — DISMISSED** [review]

File: OpenRA.Game/GameRules/ActorInfo.cs (hunk 2 above)
Comment (RoosterDragon) [reply OpenRA.Game/GameRules/ActorInfo.cs]: Updated

**Review by PunkPun — APPROVED** [review]

**Comment by PunkPun** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ca3f6b78fe3195781ed4568a16fd5b3a0d85c1a5...33b48672df31908ed6e208cea85675d9e8d24218)
